parity-clib/src/lib.rs

```diff
@@ -152,116 +196,86 @@ pub unsafe extern fn parity_rpc(client: *mut c_void, query: *const c_char, len:
 }
 
 #[no_mangle]
-pub unsafe extern fn parity_set_panic_hook(callback: extern "C" fn(*mut c_void, *const c_char, usize), param: *mut c_void) {
-	let cb = CallbackStr(Some(callback), param);
-	panic_hook::set_with(move |panic_msg| {
-		cb.call(panic_msg);
-	});
-}
+pub unsafe extern fn parity_subscribe_ws(
+	client: *const c_void,
+	query: *const c_char,
+	len: usize,
+	callback: Callback,
+	user_data: *mut c_void,
+) -> *const c_void {
 
-// Internal structure for handling callbacks that get passed a string.
-struct CallbackStr(Option<extern "C" fn(*mut c_void, *const c_char, usize)>, *mut c_void);
-unsafe impl Send for CallbackStr {}
-unsafe impl Sync for CallbackStr {}
-impl CallbackStr {
-	fn call(&self, new_chain: &str) {
-		if let Some(ref cb) = self.0 {
-			cb(self.1, new_chain.as_bytes().as_ptr() as *const _, new_chain.len())
+	panic::catch_unwind(|| {
+		if let Some((client, query)) = parity_rpc_query_checker(client, query, len) {
```

Comment:
Seems pretty similar to the previous file, can we share some logic (hopefuly doesn't need to be `unsafe`)

Reply:
Yeah, that is true. The only thing that stopped me from sharing this is how the callbacks are invoked and I didn't want to have a dependency to `jni` in the parity-clib some special types are required. However, I will double check if I can come up with a better solution.